File: in_Java/0121 Best Time to Buy and Sell Stock.java
class Solution {
    public int maxProfit(int[] prices) {
        if(prices == null || prices.length == 0) return 0;
        int curmin = prices[0];
        int res = Integer.MIN_VALUE;
        int[] t = new int[prices.length];
        for(int i = 1; i < prices.length; i++){
            if(prices[i] >= curmin){
                t[i] = prices[i] - curmin;
            }else{
                curmin = prices[i];
                t[i] = 0;
            }
        }
        for(int i : t){
            if(i > res) res = i;
        }
        return res;
    }
}

File: in_Python/0121 Best Time to Buy and Sell Stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        res = 0
        if prices == None or len(prices) == 0:
            return res
        curmin = prices[0]
        for i in range(1, len(prices)):
            p = prices[i]
            if p < curmin:
                curmin = p
            else:
                dis = p - curmin
                if dis > res:
                    res = dis
        return res